marshal.go

```diff
@@ -76,30 +100,140 @@ func (a *Application) UnmarshalJSON(b []byte) error {
-// different Port encodings between XML and JSON. Here we copy the values from the JSON
-// Port struct into the simple XML int field.
+// UnmarshalJSON is a custom JSON unmarshaler for Instance, transcribing the two composite port
+// specifications up to top-level fields.
 func (i *Instance) UnmarshalJSON(b []byte) error {
+	// Preclude recursive calls to MarshalJSON.
+	type instance Instance
+	// inboundJSONFormatPort describes an instance's network port, including whether its registrant
+	// considers the port to be enabled or disabled.
+	//
+	// Example JSON encoding:
+	//
+	//   Eureka versions 1.2.1 and prior:
+	//     "port":{"@enabled":"true", "$":"7101"}
+	//
+	//   Eureka version 1.2.2 and later:
+	//     "port":{"@enabled":"true", "$":7101}
+	//
+	// Note that later versions of Eureka write the port number as a JSON number rather than as a
+	// decimal-formatted string. We accept it as either an integer or a string. Strangely, the
+	// "@enabled" field remains a string.
+	type inboundJSONFormatPort struct {
+		Number  interface{} `json:"$"`
+		Enabled bool        `json:"@enabled,string"`
+	}
+	aux := struct {
+		*instance
+		Port       inboundJSONFormatPort `json:"port"`
+		SecurePort inboundJSONFormatPort `json:"securePort"`
+	}{
+		instance: (*instance)(i),
+	}
+	if err := json.Unmarshal(b, &aux); err != nil {
+		return err
+	}
+	resolvePort := func(port interface{}) (int, error) {
+		return intFromJSONNumberOrString(port, "port number")
+	}
 	var err error
-	var ii instance
-	if err = json.Unmarshal(b, &ii); err == nil {
-		marshalLog.Debug("Instance.UnmarshalJSON ii:%+v\n", ii)
-		*i = Instance(ii)
-		i.Port = parsePort(ii.PortJ.Number)
-		i.SecurePort = parsePort(ii.SecurePortJ.Number)
-		return nil
+	if i.Port, err = resolvePort(aux.Port.Number); err != nil {
+		return err
 	}
-	return err
+	i.PortEnabled = aux.Port.Enabled
+	if i.SecurePort, err = resolvePort(aux.SecurePort.Number); err != nil {
+		return err
+	}
+	i.SecurePortEnabled = aux.SecurePort.Enabled
+	return nil
 }
 
-func parsePort(s string) int {
-	n, err := strconv.Atoi(s)
-	if err != nil {
-		log.Warningf("Invalid port error: %s", err.Error())
+// MarshalJSON is a custom JSON marshaler for Instance, adapting the top-level raw port values to
+// the composite port specifications.
+func (i *Instance) MarshalJSON() ([]byte, error) {
+	// Preclude recursive calls to MarshalJSON.
+	type instance Instance
+	// outboundJSONFormatPort describes an instance's network port, including whether its registrant
+	// considers the port to be enabled or disabled.
+	//
+	// Example JSON encoding:
+	//
+	//   "port":{"@enabled":"true", "$":"7101"}
+	//
+	// Note that later versions of Eureka write the port number as a JSON number rather than as a
+	// decimal-formatted string. We emit the port number as a string, not knowing the Eureka
+	// server's version. Strangely, the "@enabled" field remains a string.
```

Comment:
The server doesn't send anything indicative of its version—not even in an HTTP header. If we had already received some responses from a server and been able to inspect the fields that differ, we could then remember which "style" or "version" of the encoding we've seen the server use and then mimic that in our future requests. There's still that problem of how our client starts that conversation safely.

As for whether Eureka will stop accepting the old version, I think they'd have to introduce some new resource paths in order to be fair to all the possible clients out there of various versions. My impression is that their handling of the encoding here has been haphazard; it came down to whatever [_XStream_](http://x-stream.github.io/) happened to do, and the Java clients could still talk well enough with the servers, so it doesn't appear that they cared enough to document or even prevent these format changes. I don't think they were deliberate.